data/libs/EquipSet.lua

```diff
@@ -1,404 +1,643 @@
-	listeners[self] = listeners[self] or {}
-	table.insert(listeners[self], listener)
+EquipSet.SlotTypeMatches = slotTypeMatches
+
+-- Function: CompatibleWithSlot
+--
+-- Static helper function to check if the given equipment item is compatible
+-- with the given slot object. Validates type and size parameters of the slot.
+--
+-- Parameters:
+--
+--  equip - EquipType, equipment item instance or prototype to check
+--
+--  slot  - optional HullConfig.Slot, the slot the equipment item is being
+--          validated against. If not present, the function validates that the
+--          passed equipment item is a non-slot equipment item.
+--
+---@param equip EquipType
+---@param slot HullConfig.Slot?
+function EquipSet.CompatibleWithSlot(equip, slot)
+	local equipSlot = equip.slot or false
+	if not slot then return not equipSlot end
+
+	return equipSlot and
+		slotTypeMatches(equip.slot.type, slot.type)
+		and (equip.slot.size <= slot.size)
+		and (equip.slot.size >= (slot.size_min or slot.size))
+end
+
+-- Constructor: New
+--
+-- Construct a new EquipSet object for the given ship.
+---@param ship Ship
+function EquipSet:Constructor(ship)
+	self.ship = ship
+	self.config = HullConfig.GetHullConfigs()[ship.shipId]
+
+	-- Stores a mapping of slot id -> equipment item
+	-- Non-slot equipment is stored in the array portion.
+	self.installed = {} ---@type table<string|integer, EquipType>
+	-- NOTE: the integer value stored in the cache is NOT the current array
+	-- index of the given item. It's simply a non-nil integer to indicate the
+	-- item is not installed in a slot.
+	self.cache = {} ---@type table<EquipType, string|integer>
+	-- This provides a unique index value for non-slot equipment cache entries.
+	-- It is monotonically increasing and prevents ID collisions for recursive
+	-- slots provided by non-slot equipment items.
+	self.cacheIndex = 1
+
+	-- Stores a mapping of slot id -> slot handle
+	-- Simplifies slot lookup for slots defined on equipment items
+	self.slotCache = {} ---@type table<string, HullConfig.Slot>
+	-- Stores the inverse mapping for looking up the compound id of a slot by
+	-- the slot object itself.
+	self.idCache = {} ---@type table<HullConfig.Slot, string>
+
+	self:BuildSlotCache()
+
+	-- List of listener functions to be notified of changes to the set of
+	-- installed equipment in this EquipSet
+	self.listeners = {} ---@type EquipSet.Listener[]
+
+	-- Initialize ship properties we're responsible for modifying
+	self.ship:setprop("mass_cap", self.ship["mass_cap"] or 0)
+	self.ship:setprop("equipVolume", self.ship.equipVolume or 0)
+	self.ship:setprop("totalVolume", self.ship.totalVolume or self.config.capacity)
 end
 
-function EquipSet:CallListener(slot)
-	if not listeners[self] then return end
+-- Callback: OnShipTypeChanged
+--
+-- Called when the type of the owning ship is changed. Removes all installed
+-- equipment and tries to leave the ship in an "empty" state.
+function EquipSet:OnShipTypeChanged()
```

Comment:
Q: Should it be up to the EquipSet to clear out equipment on a ship-type-changed event?

It seems to me that the ship-type should only change if the player buys a new ship, so it should be up to the Ship Marketplace to handle selling and removing the old ship (and its equipment) first before adding the new ship.

I guess it depends on the order of operations - as long as this event isn't fired until the Ship Market is finished, then having this as a final sanity-clean-up is ok.

Reply:
This is a hack. Changing ships should delete / retire the old Ship object and spawn in a new one in a "clean" state. Since we can't do that (because the player is the ship and that'd break around 30k LOC) instead we pretend we're making a new ship by getting rid of all of the equipment from the old ship. This is existing behavior that external code depends on.

Note that the ShipMarket is not the only callsite of SetShipType, but in the future it can be extended to do its own removal of the necessary equipment items, either selling them immediately or storing them to the (not yet implemented) "player's warehouse" at the station.